opengrok-indexer/src/main/java/org/opengrok/indexer/analysis/ruby/RubyLexer.java

```diff
@@ -18,128 +18,53 @@
  */
 
 /*
- * Copyright (c) 2017, Chris Fraire <cfraire@me.com>.
+ * Copyright (c) 2017, 2019, Chris Fraire <cfraire@me.com>.
  */
 
 package org.opengrok.indexer.analysis.ruby;
 
 import java.io.IOException;
 import java.util.LinkedList;
 import java.util.Queue;
+import java.util.Stack;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
+
+import org.opengrok.indexer.analysis.JFlexSymbolMatcher;
 import org.opengrok.indexer.analysis.Resettable;
 import org.opengrok.indexer.util.StringUtils;
 import org.opengrok.indexer.web.HtmlConsts;
 import org.opengrok.indexer.analysis.JFlexJointLexer;
 import org.opengrok.indexer.util.RegexUtils;
 
 /**
- * Represents an API for object's using {@link RubyLexHelper}
- */
-interface RubyLexer extends JFlexJointLexer {
-    void maybeIntraState();
-    void popHelper();
-}
-
-/**
- * Represents a helper for Ruby lexers
+ * Represents an abstract base class for Ruby lexers.
  */
-class RubyLexHelper implements Resettable {
+abstract class RubyLexer extends JFlexSymbolMatcher
+        implements JFlexJointLexer, Resettable {
 
     // Using equivalent of {Local_nextchar} from RubyProductions.lexh
     private final static Pattern HERE_TERMINATOR_MATCH = Pattern.compile(
         "^[a-zA-Z0-9_\u00160-\u0255]+");
 
-    private final RubyLexer lexer;
-
-    private final int QUOxLxN;
-    private final int QUOxN;
-    private final int QUOxL;
-    private final int QUO;
-    private final int HEREinxN;
-    private final int HERExN;
-    private final int HEREin;
-    private final int HERE;
-    private final int SCOMMENT;
-    private final int POD;
-
-    private Queue<HereDocSettings> hereSettings;
-
-    /**
-     * When matching a quoting construct like qq[], q(), m//, s```, etc., the
-     * operator name (e.g., "m" or "tr") is stored. Unlike {@code endqchar} it
-     * is not unset when the quote ends, because it is useful to indicate if
-     * quote modifier characters are expected.
-     */
-    private String qopname;
-
-    /**
-     * When matching a quoting construct like %w(), '', %[], etc., the
-     * terminating character is stored.
-     */
-    private char endqchar;
-
-    /**
-     * When matching a quoting construct like %[], %w(), %&lt;&gt; etc. that
-     * nest, the begin character ('[', '&lt;', '(', or '{') is stored so that
-     * nesting is tracked and {@code nendqchar} is incremented appropriately.
-     * Otherwise, {@code nestqchar} is set to '\0' if no nesting occurs.
-     */
-    private char nestqchar;
-
-    /**
-     * When matching a quoting construct like %[], %w(), etc., the
-     * number of remaining end separators is stored. It starts at 1, and
-     * any nesting increases the value.
-     */
-    private int nendqchar;
+    private RubyLexerData d;
```

Comment:
can this have some descriptive name?

Reply:
With variables like c, i, rem, o a simple `d` can become a candidate for a local variable, and one has to actually look for the definition and that it actually shared variable. This doesn't seem obvious for me. If that's a head of a stack, why not name it like that? `d` doesn't mean anything.